### PR TITLE
implement L4 AuthorizationPolicy

### PR DESCRIPTION
**Pull request by howardjohn**
This commit introduces the first iteration of L4 AuthorizationPolicy
support.

* Policies are read over XDS and converted to a new internal type. These
  are indexed by namespace
* Inbound requests check against applicable policies and send 401 in the
  HBONE response when they fail to match
* Current matching algorithm is fairly efficient (TODO: benchmarks), but
  not optimized (one could imagine a trie like implementation; today it
is a linear search).

Fixes https://github.com/istio/ztunnel/issues/259
Fixes https://github.com/istio/ztunnel/issues/205
Fixes https://github.com/istio/ztunnel/issues/29

### CODE REVIEW

File: src/proxy/inbound.rs [hunk 1]
```diff
@@ -73,8 +76,25 @@ impl Inbound {
...
+                src_identity: socket
+                    .ssl()
+                    .peer_certificate()
+                    .and_then(|x| crate::tls::boring::extract_sans(&x).first().cloned()),
+                src_ip: to_canonical(socket.get_ref().peer_addr().unwrap()).ip(),
+                dst,
+            };
+            let workloads = self.workloads.clone();
+            async move {
+                let conn = conn.clone();
```
Comment (stevenctl) [inline src/proxy/inbound.rs:91]: redundant clones?

File: src/proxy/inbound.rs [hunk 2]
```diff
@@ -187,6 +215,22 @@ impl Inbound {
...
+                if addr.ip() != conn.dst.ip() {
+                    info!("Sending 400, ip mismatch {addr} != {}", conn.dst);
+                    return Ok(Response::builder()
+                        .status(StatusCode::BAD_REQUEST)
+                        .body(Body::empty())
+                        .unwrap());
+                }
+                // Orig has 15008, swap with the real port
+                let conn = rbac::Connection { dst: addr, ..conn };
+                if !workloads.assert_rbac(&conn).await {
```
Comment (stevenctl) [inline src/proxy/inbound.rs:227]: Just tested locally with the local redirect method and I'm getting:

```
2022-12-28T16:48:39.049473Z  INFO inbound{id=2f63be3786443ac0803b0dd43a307b6a peer_ip=127.0.0.1 peer_id=spiffe://cluster.local/ns/default/sa/default}: ztunnel::proxy::inbound: RBAC rejected conn=127.0.0.1(spiffe://cluster.local/ns/default/sa/default)->127.0.0.1:8000
```

Comment (stevenctl) [reply src/proxy/inbound.rs:227]: `no allow policies matched`

Comment (howardjohn) [reply src/proxy/inbound.rs:227]: do you have policies defined?

File: src/rbac.rs [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+        } else {
+            let matches = positive.iter().any(&mut predicate);
+            trace!(%matches, "type"="positive", "{positive:?}");
+            matches
+        };
+        let nm = if negative.is_empty() {
+            trace!(matches = true, "type" = "negative", "no match declared");
+            true
+        } else {
+            let matches = negative.iter().any(&mut predicate);
```
Comment (stevenctl) [inline src/rbac.rs:169]: Negate predicate this or change `any` to `none`

File: src/proxy/inbound.rs (hunk 2 above)
Comment (stevenctl) [reply src/proxy/inbound.rs:227]: There is an `allow: not 9999` in local.yaml but I don't think it's matching correctly. See the other comment.

File: src/rbac.rs (hunk 3 above)
Comment (stevenctl) [reply src/rbac.rs:169]: or I guess just make it check `pm && !nm` is cleaner

Comment (howardjohn) [reply src/rbac.rs:169]: good catch - will add a bunch of tests before we merge this

**Review by stevenctl — APPROVED** [review]

File: proto/networkpolicy.proto [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  //  MatchBehavior match_behavior = 1;
+  repeated RBACPolicyRuleMatch matches = 2;
+}
+
+message RBACPolicyRuleMatch {
+  // Values of specific type are ORed
+  // If multiple types are set, they are ANDed
+
+  repeated StringMatch namespaces = 1;
+  repeated StringMatch not_namespaces = 2;
```
Comment (hzxuzhonghu) [inline proto/networkpolicy.proto:56]: No need using stringMatch for namespaces, makes things a little more complex

File: proto/networkpolicy.proto [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    string suffix = 3;
+
+    google.protobuf.Empty presence = 4;
+  }
+}
+
+enum RBACScope {
+  // ALL means that the authorization policy will be applied to all workloads
+  // in the mesh (any namespace).
+  GLOBAL = 0;
```
Comment (hzxuzhonghu) [inline proto/networkpolicy.proto:93]: Whyt not use Namespace="" to represent all namespaces

File: proto/networkpolicy.proto (hunk 4 above)
Comment (howardjohn) [reply proto/networkpolicy.proto:56]: It's part of AurhorizationPolicy unfortunately. I agree but we can't really change it at this point

File: proto/networkpolicy.proto [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+  // NAMESPACE means that the policy will only be applied to workloads in a
+  // specific namespace.
+  NAMESPACE = 1;
+  // WORKLOAD_SELECTOR means that the policy will only be applied to specific
+  // workloads that were selected by their labels.
+  WORKLOAD_SELECTOR = 2;
```
Comment (hzxuzhonghu) [inline proto/networkpolicy.proto:99]: Cross namespaces selector seems very weird

File: proto/networkpolicy.proto (hunk 5 above)
Comment (howardjohn) [reply proto/networkpolicy.proto:93]: It's still useful to know which namespace it came from even if it's a global rule, IMO

File: proto/networkpolicy.proto (hunk 6 above)
Comment (howardjohn) [reply proto/networkpolicy.proto:99]: It's part of the authz v1 API unfortunately

**Pull request merged**